### ISSUE TITLE
Remove or neuter DeleteActor syscall

### ISSUE THREAD

**Issue by anorth** [filecoin-project/fvm-specs#100]
I think we should remove the DeleteActor syscall. Or, if something actively prevents that, reduce its functionality to leave the top-level actor object in the state tree. The critical state to preserve is the nonce.

## Why do we have DeleteActor?
Chain state is expensive, and so removing it sounds like a good idea. Filecoin was inspired by Ethereum here, which has gone through a similar journey with SELFDESTRUCT.
- The call was conceived as having a negative gas cost, thus incentivising the removal of state. In Ethereum this turned out to have  the counterproductive effect of incentivising a gas futures marketplace which stored _more_ state, so that some of it could be deleted later in time when gas (for computation) was expensive.
- [EIP-3529](https://eips.ethereum.org/EIPS/eip-3529) removed the refund, and the FVM follows suit. So now there is no incentive to SELFDESTRUCT – it would make code larger and be purely altruistic.

DeleteActor has one uses in the built-in actors.
- A payment channel will delete itself after being collected

As a built-in actor, the forced "altruism" of deletion makes some sense. But we should expect in the future that new and better user-programmed payment channels will dominate usage, and they will not pay the extra cost.

(A previous use case where a defunct miner actor was deleted has since been removed.)

## Why should we remove it?
At this point, pre-FVM launch, I think the appropriate question should be "why should we retain it?", with a default expectation of removing any complexity that's not well motivated. I don't know of a good reason: it's very rarely used in Ethereum, and the patterns it's used for (e.g. a faux multicall) usually have a more natural expression that we could support instead.

DeleteActor adds complexity because it changes the possible states for an address to go from no actor -> actor then to also include actor -> no actor, and then to cycle. With [predictable actor address generation](https://github.com/filecoin-project/FIPs/discussions/379), an address could be re-used. While we might (with careful thought) be able to ensure that only the same code is deployed to that address, the actor nonce would recycle, opening up replay attacks for smart-contract wallets with account abstraction. One could not generally rely on any monotonicity of state implied by the code. This would also break any assumption that a message CID is can be included only once in the chain.

If we later build in a contract upgrade mechanism (to avoid proxy contracts), self-deleting actors will also complicate reasoning about the address/s authorized to deploy to the actor in question.

DeleteActor can unlink an arbitrarily large amount of state. While a net benefit in the long term, it's not clear that this is a "free" operation for nodes to perform. It would certainly complicate caching approaches. (We can unlink arbitrarily large amounts of state anyway, though).

There is strong support in Ethereum-land for removing or neutering their version of `SELFDESTRUCT` for similar reasons. [EIP-4758](https://eips.ethereum.org/EIPS/eip-4758) renames it to `SENDALL` which transfers balance but does not delete any state.  This is [not the first](https://github.com/ethereum/EIPs/pull/2751) proposal, but the proximate motivation this time is that deleting arbitrary storage is incompatible with Verkle trees as a state representation. 

Who can tell what other future innovation would conflict with actor deletion? The feature doesn't bring any benefits to justify the complexity.

## Alternatives
I think just removing the syscall, and it's one use in payment channels, is the simplest and easiest to think about approach.

If we don't do that, I think we must neuter it to leave the top-level actor object (including nonce) in place, which would prevent re-use of the address. We _could_ clear the Code CID and State CID.

**Comment by Stebalien** [comment]
> Why do we have DeleteActor?

An additional motivation is security. For example, if an actor is an "owner" of another actor, it would be nice to have a clear way to "neuter" the actor.

However, this can also be done by just setting the actor into some "dead" state.

> an address could be re-used

I don't think this is the case for us:

1. The address in the state-tree is an ID address, and we assign ID addresses in ascending order.
2. We don't remove the actor from the _init_ actor, so we can't reuse those addresses.

So yes, I think we should have a strong "no address may be reused" guarantee, but I'd like the ability to at least "tombstone" an actor in such a way that it's easy to verify that the actor can never be revived.

**Comment by anorth** [comment]
I agree that IDs won't be re-used. But if we allow full deletion and predictable addresses, then a predictable address could be re-used and be assigned a new, different ID than it had last time!

**Comment by Stebalien** [comment]
> I agree that IDs won't be re-used. But if we allow full deletion and predictable addresses, then a predictable address could be re-used and be assigned a new, different ID than it had last time!

Ah, my point is that we currently don't remove addresses, even if we delete the actor. It's perfectly valid to have an address pointing to a non-existent actor.

(steb now goes to check that this isn't a fatal error in the FVM...)
